crypto/provider_core.c

```diff
@@ -1672,6 +1674,28 @@ static int core_pop_error_to_mark(const OSSL_CORE_HANDLE *handle)
 {
     return ERR_pop_to_mark();
 }
+
+static int core_obj_add_sigid(const OSSL_CORE_HANDLE *prov,
+                              const char *sign_name, const char *digest_name,
+                              const char *pkey_name)
+{
+    int sign_nid = OBJ_txt2nid(sign_name);
+    int digest_nid = OBJ_txt2nid(digest_name);
+    int pkey_nid = OBJ_txt2nid(pkey_name);
+
+    if (sign_nid == NID_undef
+            || digest_nid == NID_undef
+            || pkey_nid == NID_undef)
+        return 0;
+
+    return OBJ_add_sigid(sign_nid, digest_nid, pkey_nid);
+}
+
+static int core_obj_create(const OSSL_CORE_HANDLE *prov, const char *oid,
+                           const char *sn, const char *ln)
+{
+    return (OBJ_create(oid, sn, ln) != NID_undef);
```

Comment:
Should there be graceful return if the sigid or the individual ids are already inserted? Otherwise the providers would simply have to ignore errors from these calls when loaded/unloaded multiple times.

Reply:
The problem is that the provider author does not really know what the error is in that case. I would be OK with just explicitly documenting that the error might mean the ids were added already and so it should be just ignored and further calls skipped.

Reply:
If we want to mimic the OBJ_create() behaviour, we should have an up-call to query an OID.  This would allow a provider to determine if it needs to create the OID or not.

The alternative would be @t8m's suggestion of making these safer.  Could they return success if the OID already exists?  The provider wants to create the OID, if it already exists it's as good as creating it -- it's not like there is anything else we're letting the provider do with the OID (at least at this stage).

Reply:
I've amended it so that the function returns success if it successfully created the object - or if the OID already exists.